quarkus/addons/token-exchange/runtime/src/main/java/org/kie/kogito/addons/quarkus/token/exchange/cache/TokenPolicyManager.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.kie.kogito.addons.quarkus.token.exchange.cache;
+
+import java.util.concurrent.TimeUnit;
+
+import org.kie.kogito.addons.quarkus.token.exchange.utils.CacheUtils;
+import org.kie.kogito.addons.quarkus.token.exchange.utils.ConfigReaderUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.Expiry;
+
+public class TokenPolicyManager {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TokenPolicyManager.class);
+
+    /**
+     * Creates an expiry policy that uses each token's actual expiration time
+     */
+    public static Expiry<String, CachedTokens> createTokenExpiryPolicy() {
+        return new Expiry<String, CachedTokens>() {
+            @Override
+            public long expireAfterCreate(String key, CachedTokens value, long currentTime) {
+                return calculateTimeToExpiration(key, value);
+            }
+
+            @Override
+            public long expireAfterUpdate(String key, CachedTokens value, long currentTime, long currentDuration) {
+                return calculateTimeToExpiration(key, value);
+            }
+
+            @Override
+            public long expireAfterRead(String key, CachedTokens value, long currentTime, long currentDuration) {
+                return currentDuration; // Don't change expiration on read
+            }
+        };
+    }
+
+    /**
+     * Calculate time to expiration based on token's actual expiration time minus proactive refresh buffer
+     */
+    private static long calculateTimeToExpiration(String cacheKey, CachedTokens tokens) {
+        String authName = CacheUtils.extractAuthNameFromCacheKey(cacheKey);
+        long proactiveRefreshSeconds = ConfigReaderUtils.getProactiveRefreshSeconds(authName);
```

Comment:
You could also add this limit to make it more robust and avoid incorrect configuration (negative numbers)
```suggestion
        long proactiveRefreshSeconds = Math.max(0, ConfigReaderUtils.getProactiveRefreshSeconds(authName));
```